src/reactor.rs

```diff
@@ -403,9 +407,28 @@ mod sys {
             epoll_ctl(self.0, EpollOp::EpollCtlDel, fd, None).map_err(io_err)
         }
         pub fn wait(&self, events: &mut Events, timeout: Option<Duration>) -> io::Result<usize> {
-            let timeout_ms = timeout
-                .and_then(|t| t.as_millis().try_into().ok())
-                .unwrap_or(-1);
+            let timeout_ms = match timeout {
+                // If Duration is zero, there's no need for timerfd. (also because it means disarming timer)
+                Some(t) if t.as_nanos() == 0 => 0,
+                // Otherwise, set timerfd with nanosecond precision
+                Some(t) => unsafe {
+                    static TFD: Lazy<i32> =
+                        Lazy::new(|| unsafe { timerfd_create(CLOCK_REALTIME, TFD_CLOEXEC) });
+                    let timespec = *TimeSpec::nanoseconds(t.as_nanos() as i64).as_ref();
+                    let new_value = itimerspec {
+                        it_interval: timespec,
+                        it_value: timespec,
+                    };
+                    // Start a relative timer
+                    let res = timerfd_settime(*TFD, 0, &new_value, std::ptr::null_mut());
+                    Errno::result(res).map(|r| r as usize).map_err(io_err)?;
+                    // TFD_NONBLOCK isn't set, so this will wait for the first timeout.
```

Comment:
Why would you do it like that instead of registering timerfd with epoll? Seems like it would block any other events.

Reply:
If it is registered to epoll, doesn't mean the precision is still millisecond? And isn't this method the reason to block I/O events? (https://github.com/stjepang/smol/blob/master/src/reactor.rs#L261)

Reply:
>If it is registered to epoll, doesn't it mean the precision is still millisecond?

No. The kernel will return from `epoll_wait` as soon as the timer fires if it is in the epoll set.

>And isn't this method the reason to block I/O events?

Yes, it blocks waiting on *multiple* i/o sources waiting for events. You block on one source (timer), which means that other events won't be processed until timer fires.

Reply:
Ah thanks for clarify. I think I'll close it for now to fix this and also the android target issue. Will reopen it once everything is resolved.